---
title: "ddirch_covariance_problem"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Microbial amplicon counts, relative abundances, and covariance.

Microbial communities are often characterizied by sequencing marker gene regions that can then be binned into oeprational taxonomic units (OTUs), which we think about as species. Data is returned as a species matrix. Each sample is a vector of counts, and the length of the vector is the total number of species observed across all samples. The values within that vector correpsond to the number of times each OTU was observed in that sample (or the number of amplicons per OTU. An amplicon is just a single read of an amplified PCR product. The PCR product in this case is the marker gene region of interest). Here are some example data as a species/OTU matrix. The columns are unique species. The rows are the abundances of those species in each sample, with each row repesenting a unique sample.
```{r}
   y1 <- round(rnorm(100, 100, 5)) 
   y2 <- round(y1*-2 + rnorm(100,0,5)) + 400
   y3 <- round(rnorm(100,50,5))
spp.y <- as.matrix(data.frame(y1,y2,y3))
```

In this example species `y1` and `y2` are negatively correlated with eachother. Species `y3` is not correlated with the abundances of species `y1` or `y2`. Here is a visualization of this:

```{r, echo = F}
pairs(spp.y)
```

## Microbial amplicon data are compositional. This matters.
The total number of reads (or amplicons or sequences or observations) per sample varies from sample to sample due to the sequencing isntrument. One sample may have been sampled to a sequencing depth of 1,000 and the next to 10,000. If both of these samples contain 50% of species 1 and 50% of species 2, then sample 1 will have 500 reads of each species, while sample 2 will have 5,000 reads of each species. The total number of reads doesn't matter, just the relative abundances. As a result, we must transform our species matrix into relative abundances.

```{r}
rel.spp.y <- spp.y/rowSums(spp.y)
```

This is where problems start to arise. Because the data are compositional, straight-forward regression analysis goes out the window when attempting to understand the relationships *among* species (i.e. is species 1 positively or negatively correlated with species 2?) This can be understood pretty quickly by looking at a correlation plot of the relative abundances.

```{r, echo = F}
pairs(rel.spp.y)
```

There is now a significant negative correlation between species 2 and species 3, even though in reality there is no correlation in the abundances of these species. They were drawn randomly from independent distributions above. This gets to the core of the question we want to answer: **How can we quantify covariance among species abundances in our compositional data?**

## Fitting species abundances using a dirlichet distribution
Before addressing the covariance issue we wanted to first fit an intercept-only model to the abundances of each species, without accounting for covariance among species abundances. The dirlichet distribution is the multivariate generalization of the beta distribution, and allows us to model the abundances of all species simultaneously, while accounting for the fact that relative abundances of all species witih a sample must sum to 1. We fit this model in JAGS, implemented in R.

```{r}
jags.model = "
model {
    #setup priors for each species
    for(j in 1:N.spp){
      m0[j] ~ dgamma(1.0E-3, 1.0E-3) #intercept prior
    }

    #go ahead and fit means of species abundances.
    for(i in 1:N){
        for(j in 1:N.spp){
             log(a0[i,j]) <- m0[j] 
           }
     
    y[i,1:N.spp] ~ ddirch(a0[i,1:N.spp]) 
    }

} #close model loop.
"
```

Go ahead and setup a JAGS data object and fit this model (it runs quickly), see if it can capture the true relative abundances.

```{r, message = F}
library(runjags)
#JAGS data object as list.
jags.data <- list(    y = rel.spp.y,
                      N = nrow(rel.spp.y),
                  N.spp = ncol(rel.spp.y))

#Fit the JAGS model.
jags.out <- run.jags(jags.model,
                       data = jags.data,
                      adapt = 100,
                     burnin = 200,
                     sample = 400,
                   n.chains = 3,
                    monitor = c('m0'))
#summarize output (get distributions on m0 values, the intercepts)
mod.sum <- summary(jags.out)
```

Modeled relative abundances:
```{r}
#check if fitted relative abundances match true relative abundances.
#fitted
a.m <- exp(summary(jags.out)[,"Median"])
a.m / sum(a.m)
```

True relative abundances:
```{r}
#actual
colMeans(rel.spp.y)
```

The model does a great job of fitting the mean species relative abundances across samples. cool. Its fairly easy to generalize this to add more predictors where `log(a0[i,j])` is defined in the model.

## But what about the covariance??

This is where things get hard. Again, just looking at correlations among species/OTU relative abundances is inappropriate, because you will quickly get excited about strong negative correlations that are total artifacts (see the spurrious correlation between species 2 and species 3 above.) People in the microbiome literature have though about this a lot. There is a nice summary in this article: [Microbiome Datasets Are Compositional: And This is Not Optional](https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full). There are also methods that have been developed to deal with this problem, such as [SparCC](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002687) or [SPIEC-EASI](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004226). However, both of these methods require the data being sparse. This is common in microbiome studies, as you generally have 1000s of species/OTUs, and many samples have lots of zero observations for OTU abundances. This is precisely *not* our case, as when we bin OTUs into functional groups we reduce the effective diversity of the sample from 1000s to a number less than ten. Our data are not sparse. These methods generally take advantage of the fact that although the relative abundance of species within a sample is not indicative of the true abundances, the *proportion* of relative abundances of two species is equal to the proportion of the same two species' true abundances in that sample. Then, by analyzing proportions, they can get somewhere.

## What about a multivariate normal that accounts for variation in total number of sequences?
I figured one way to get around the compositional problem is to convert it back to true abundances. The variation in actual sequencing depth is actually more extreme than the simulated data above, but I figured if this approach works, then this should be an easy case.

#### First I define a multivariate normal JAGS model with covariance.
```{r}
jags.model = "
model {
#setup priors for each species
for(j in 1:N.spp){
  m0[j] ~ dnorm(0, .001) #intercept prior
}

#Sigma (covariance matrix) prior
sigma ~ dwish(r,S)

#implement multivariate normal.
for(i in 1:N){
  y.hat[i,1:N.spp] <- (m0[1:N.spp]) * seqs[i]
      y[i,1:N.spp] ~ dmnorm(y.hat[i,1:N.spp], sigma[1:N.spp,1:N.spp])
}

} #close model loop.
"
```

#### then I give a wishart prior on the covariance matrix that is super weak.
```{r}
W = diag(ncol(spp.y))/10000
```

#### go ahead and define JAGS data object and fit the model using original spp.y data without relativea abundance transform.

```{r, message = F}
jags.data <- list(y = spp.y,
                  N = nrow(spp.y),
                  N.spp = ncol(spp.y),
                  r = W,
                  S = ncol(spp.y),
                  seqs = rowSums(spp.y))

jags.out <- run.jags(jags.model,
                     data=jags.data,
                     adapt = 100,
                     burnin = 3000,
                     sample = 2000,
                     n.chains=3,
                     monitor=c('m0','sigma'))
```

#### check that this model captures the true relative abundances (it does).
true relative abundances
```{r}
colMeans(spp.y) / sum(colMeans(spp.y)) 
```

estimated relative abundances
```{r}
summary(jags.out)[grep('m0',rownames(summary(jags.out))),'Median']
```

#### Check the covariance matrices.
true covariance matrix
```{r}
cov(spp.y)
```

estimated covariance matrix
```{r}
z <- summary(jags.out)[grep('sigma',rownames(summary(jags.out))),'Median']
i.covm <- matrix(z, ncol = sqrt(length(z)))
solve(i.covm) #JAGS returns the inverse of the covariance matrix. This inverts it.
```

These clearly do not match. Still seems to be a spurrious covariance between species 2 and species 3.

## What to try next?
Mike has encouraged me to look into the generalized dirlichet distribution, however this is getting a bit over my head, as its challenging for me to read the advanced primary statistics literature. Right now we're moving forward modeling the relative abundances using the standard dirlichet distribution, however we'd really like to get at this covariance problem, as it would inform our understanding of microbial species or functional group interactions.